File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        
        backtrack("(", 1, 0, res, n);
        return res;
    }
    
    static void backtrack(String curr, int op, int cl, List<String> res, int n) {
        
        if (curr.length() == 2*n) {
                 res.add(curr);
              return;
        }
        
        if (op < n) backtrack(curr + "(", op + 1, cl, res, n);
        if (cl < op) backtrack(curr + ")", op, cl + 1, res, n);
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        
        def backtrack(openB, closedB, currstr):
            if openB == closedB == 0:
                res.append(currstr)
                return 
            
            elif closedB < openB or openB < 0 or openB > n:
                return 
            
            backtrack(openB - 1, closedB, currstr + "(")
            backtrack(openB, closedB - 1, currstr + ")")
            
            
        backtrack(n, n, "")
        return res
        